### PR TITLE
Fix recursion error when fetching auth chain over federation

### PR DESCRIPTION
**Pull request by erikjohnston**
When fetching the state of a room over federation we receive the event
IDs of the state and auth chain. We then fetch those events that we
don't already have.

However, we used a function that recursively fetched any missing auth
events for the fetched events, which can lead to a lot of recursion if
the server is missing most of the auth chain. This work is entirely
pointless because would have queued up the missing events in the auth
chain to be fetched already.

Let's just diable the recursion, since it only gets called from one
place anyway.

### CODE REVIEW

File: synapse/handlers/federation.py [hunk 1]
```diff
@@ -618,6 +618,10 @@ async def _get_events_from_store_or_dest(
         will be omitted from the result. Likewise, any events which turn out not to
         be in the given room.
 
+        This function *does not* recursively get missing auth events of the
+        newly fetched events. Callers must include any missing events from the
```
Comment (richvdh) [inline synapse/handlers/federation.py:622]: must include where?

File: synapse/handlers/federation.py [hunk 2]
```diff
@@ -1150,17 +1158,7 @@ async def get_event(event_id: str):
...
-                        destination, room_id, event.auth_event_ids()
-                    )
-                    auth = {}
-                    for auth_event_id in event.auth_event_ids():
-                        ae = auth_events.get(auth_event_id)
-                        if ae:
-                            auth[(ae.type, ae.state_key)] = ae
-
-                    event_infos.append(_NewEventInfo(event, None, auth))
+                    event_map[event.event_id]
```
Comment (richvdh) [inline synapse/handlers/federation.py:1161]: this looks like a no-op

File: synapse/handlers/federation.py [hunk 3]
```diff
@@ -1172,6 +1170,25 @@ async def get_event(event_id: str):
 
         await concurrently_execute(get_event, events, 5)
 
+        # Make a map of auth events for each event. We do this after fetching
+        # all the events as some of the events auth events will be in the list
```
Comment (babolivier) [inline synapse/handlers/federation.py:1174]: Either there's an apostrophe missing or an extra "events" here

File: synapse/handlers/federation.py (hunk 2 above)
Comment (babolivier) [inline synapse/handlers/federation.py:1161]:
```suggestion
                    event_map[event.event_id] = event
```

**Comment by erikjohnston** [comment]
Tests are fixed by https://github.com/matrix-org/sytest/pull/911

File: synapse/handlers/federation.py [hunk 4]
```diff
@@ -618,6 +618,10 @@ async def _get_events_from_store_or_dest(
         will be omitted from the result. Likewise, any events which turn out not to
         be in the given room.
 
+        This function *does not* recursively get missing auth events of the
+        newly fetched events. Callers must include in the `event_ids` argument
+        any missing events from the auth chain.
```
Comment (richvdh) [inline synapse/handlers/federation.py:623]: not just missing events

```suggestion
        This function *does not* automatically get missing auth events of the
        newly fetched events. Callers must include the full auth chain of
        of the missing events in the `event_ids` argument, to ensure that any
        missing auth events are correctly fetched.
```

File: synapse/handlers/federation.py [hunk 5]
```diff
@@ -1131,12 +1135,16 @@ async def _get_events_and_persist(
     ):
         """Fetch the given events from a server, and persist them as outliers.
 
+        This function *does not* recursively get missing auth events of the
+        newly fetched events. Callers must include in the `event_ids` argument
```
Comment (richvdh) [inline synapse/handlers/federation.py:1139]:
```suggestion
        newly fetched events. Callers must include in the `events` argument
```

File: synapse/event_auth.py [hunk 6]
```diff
@@ -69,10 +69,11 @@ def check(
     # sanity-check it
     for auth_event in auth_events.values():
         if auth_event.room_id != room_id:
-            raise Exception(
+            raise AuthError(
+                500,
```
Comment (richvdh) [inline synapse/event_auth.py:73]: shouldn't this be a 403?

Comment (richvdh) [reply synapse/event_auth.py:73]: oh oof there is AuthError(500) above. whatever.

Comment (erikjohnston) [reply synapse/event_auth.py:73]: Oh, yes, probably actually. I was copying it from a previous line, but I don't think that makes sense

Comment (richvdh) [reply synapse/event_auth.py:73]: but also: if we actually hit this code and it matters, the comment needs changing.

Comment (richvdh) [reply synapse/event_auth.py:73]: /me awards past me a prize for including sanity checks for security-critical things

Comment (erikjohnston) [reply synapse/event_auth.py:73]: Sorry, which comment?

Comment (erikjohnston) [reply synapse/event_auth.py]: Found it

**Pull request merged**